eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala

```diff
@@ -781,51 +808,62 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
-              log.warning("cannot do splice: invalid splice-out script")
-              cmd.replyTo ! RES_FAILURE(cmd, InvalidSpliceRequest(d.channelId))
-              stay()
+      if (d.commitments.params.remoteParams.initFeatures.hasFeature(SplicePrototype)) {
+        if (d.commitments.params.useQuiescence && d.spliceStatus == SpliceStatus.NoSplice) {
+          startSingleTimer(QuiescenceTimeout.toString, QuiescenceTimeout(peer), nodeParams.channelConf.revocationTimeout)
+          if (d.commitments.changes.localChanges.all.isEmpty) {
+            stay() using d.copy(spliceStatus = SpliceStatus.InitiatorQuiescent(cmd)) sending Stfu(d.channelId, 1)
+          } else {
+            stay() using d.copy(spliceStatus = SpliceStatus.QuiescenceRequested(cmd))
+          }
+        } else {
+          handleNewSplice(cmd, d)
+        }
+      } else {
+        log.warning("cannot initiate splice, peer doesn't support splices")
+        cmd.replyTo ! RES_FAILURE(cmd, CommandUnavailableInThisState(d.channelId, "splice", NORMAL))
+        stay()
+      }
+
+    case Event(msg: Stfu, d: DATA_NORMAL) =>
+      if (d.commitments.params.useQuiescence) {
+        d.spliceStatus match {
+          case SpliceStatus.NoSplice =>
+            startSingleTimer(QuiescenceTimeout.toString, QuiescenceTimeout(peer), nodeParams.channelConf.quiescenceTimeout)
+            if (d.commitments.latest.changes.localChanges.all.isEmpty) {
+              stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent) sending Stfu(d.channelId, 0)
             } else {
-              log.info(s"initiating splice with local.in.amount=${cmd.additionalLocalFunding} local.in.push=${cmd.pushAmount} local.out.amount=${cmd.spliceOut_opt.map(_.amount).sum}")
-              val spliceInit = SpliceInit(d.channelId,
-                fundingContribution = fundingContribution,
-                lockTime = nodeParams.currentBlockHeight.toLong,
-                feerate = targetFeerate,
-                fundingPubKey = keyManager.fundingPublicKey(d.commitments.params.localParams.fundingKeyPath, parentCommitment.fundingTxIndex + 1).publicKey,
-                pushAmount = cmd.pushAmount,
-                requireConfirmedInputs = nodeParams.channelConf.requireConfirmedInputsForDualFunding
-              )
-              stay() using d.copy(spliceStatus = SpliceStatus.SpliceRequested(cmd, spliceInit)) sending spliceInit
+              stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
             }
-          } else {
-            log.warning("cannot initiate splice, channel is not idle or peer doesn't support splices")
-            cmd.replyTo ! RES_FAILURE(cmd, CommandUnavailableInThisState(d.channelId, "splice", NORMAL))
+          case SpliceStatus.QuiescenceRequested(_) =>
+            stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
```

Comment:
This case is a red herring that we're not handling correctly the race between our `stfu` and theirs, because we're switching from initiator to non-initiator implicitly here (which would create a deadlock if we were the channel initiator).

But there is an issue in the data model, which currently doesn't let us do that easily, because if we're the channel initiator and stay in `QuiescenceRequested`, we lose the information that our peer has already sent us `stfu` and will deadlock later waiting for them to send it.

We probably need to add a field `remoteStfuReceived: Boolean` to `QuiescenceRequested` to handle that case?

Reply:
I don't see how we can get a deadlock here. There should be no race possible between our `stfu` and theirs in the `SpliceStatus.QuiescenceRequested` state because we have not sent an `stfu` yet. This state means we are simply waiting for our local pending updates to finalize before sending an initial `stfu`.

If a node in this state receives `stfu` from it's peer before it has sent it's own `stfu`, then it's safe to switch to being the non-initiator because we have not sent our own `stfu` yet.

Reply:
Good catch, you're right, this works fine!